pingcap/tidb
# planner: fix oom by LRUPlanCache's key expand

### What problem does this PR solve?


Issue Number: close #38244 

Problem Summary:
The plan cache key change with `schemaversion` change, when evict the expired plan I did't delete the expired key causing the memory expand
### What is changed and how it works?
Remove the expired key in LRU
### Check List

Tests 


Side effects


Documentation


### Release note



Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```
### Manual test (detailed scripts )
* Set plancache with default size(100).
* Use sysbench do oltp_write_only.
* Use a Python script add/drop column in table `sbtest1` all the time, see the bucket size in debug mode.
```
# sysbench command 
sysbench --mysql-user=root --db-driver=mysql --time=0 --report-interval=10 --mysql-db=sysbench --mysql-host=127.0.0.1  --mysql-port=4000 --threads=8 --tables=4 --table-size=10000 --mysql-ignore-errors=2013,1213,1105,1205,8022,8027,8028,9004,9007 --rand-type=special oltp_write_only 
```
Before:
![飞书20220930-150304](https://user-images.githubusercontent.com/108643977/193211897-e6fd666b-2d5b-4898-af12-746499b7ed2d.jpg)
After:
![飞书20220930-150307](https://user-images.githubusercontent.com/108643977/193211931-b5472abd-e854-470d-bf3a-dc7bee53f659.jpg)
